src/drivers/uavcan/node_info.cpp

```diff
@@ -0,0 +1,311 @@
+
+	startTimerIfNotRunning();
+}
+
+void NodeInfoPublisher::handleNodeInfoUnavailable(uavcan::NodeID node_id)
+{
+}
+
+void NodeInfoPublisher::handleTimerEvent(const uavcan::TimerEvent &event)
+{
+	// Publish device information using round-robin approach
+	publishDeviceInformationPeriodic();
+}
+
+void NodeInfoPublisher::startTimerIfNotRunning()
+{
+	if (!TimerBase::isRunning()) {
+		TimerBase::startPeriodic(uavcan::MonotonicDuration::fromMSec(DEVICE_INFO_PUBLISH_INTERVAL_MS));
+	}
+}
+
+void NodeInfoPublisher::registerDevice(uint8_t node_id, const NodeInfo *info, uint32_t device_id,
+				       DeviceCapability capability)
+{
+	const bool is_registering_info = (info != nullptr);
+
+	DeviceInformation multi_capability_info;
+
+	for (size_t i = 0; i < _device_informations_size; ++i) {
+		if (is_registering_info) {
+			// Case 1: Check if this entry already has node info - skip this specific entry
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].has_node_info) {
+
+				continue;  // Continue to check other entries with same node_id
+			}
+
+			// Case 2: Check if node_id already exists with capability but no info - update that entry
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].capability != DeviceCapability::NONE &&
+			    !_device_informations[i].has_node_info) {
+				populateDeviceInfoFields(_device_informations[i], *info);
+				publishSingleDeviceInformation(_device_informations[i]);
+				continue;
+			}
+
+		} else { // registering capabilities
+			// Case 1: Check if this exact capability already exists - skip
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].device_id == device_id &&
+			    _device_informations[i].capability == capability) {
+				return;
+			}
+
+			// Case 1b: if this node has multiple capabilities, continue
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].capability != DeviceCapability::NONE  &&
+			    _device_informations[i].capability != capability) {
+				multi_capability_info = _device_informations[i];
+				continue;
+			}
+
+			// Case 2: Check if node_id already exists with node info but no capability - update that entry
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].has_node_info &&
+			    _device_informations[i].capability == DeviceCapability::NONE) {
+				_device_informations[i].device_id = device_id;
+				_device_informations[i].capability = capability;
+				publishSingleDeviceInformation(_device_informations[i]);
+				return;
+			}
+		}
+	}
+
+
+
+	// Case 3: extend array and add entry at the end
+	if (extendDeviceInformationsArray()) {
+		_device_informations[_device_informations_size - 1] = DeviceInformation();
+		_device_informations[_device_informations_size - 1].node_id = node_id;
+
+		if (multi_capability_info.serial_number[0] != '\0' && multi_capability_info.firmware_version[0] != '\0') {
```

Comment:
This is a bit fragile. What if the device does not provide the firmware version?
Using an index would be better here too.